### DISCUSSION TITLE
How can I disable the Language Model during conformer-transformer ASR training/test?

### DISCUSSION THREAD

**Discussion by ccyang123** [speechbrain/speechbrain#1828] [Q&A]
I am using the conformer-transformer ASR model with LibriSpeech dataset to perform the ASR model training and test.

The default setting of conformer-transformer training/evaluation seems to include transformer LM, 
How I can disable the Language Model during conformer-transformer ASR training/test?

**Comment by ccyang123** [comment]
If I set  the "**lm_weight = 0**" in transformer.yaml, does it mean I disable the LM model?
Thank you!

**Reply by anautsch** [reply]
Hey @ccyang123 !

If you set `lm_weight = 0` in the yaml, it is only affecting a weighted sum computation. Basically, the TransformerLM will compute all and then its score will be nullified. The computation remains, and that w/o impact.

To get rid off it, you need to get the TransformerLM entirely out of the yaml & recipe files. Yet, you might need some form of scoring unit? There are also greedy decoders/searchers, as in the wav2vec2 pretrained models, compare: https://huggingface.co/speechbrain/asr-wav2vec2-librispeech/blob/main/hyperparams.yaml

Hm. What are you up to? (If I may ask.)

**Reply by ccyang123** [reply]
Hello, @anautsch :

Thank you for your reply.
What I would like to do is to compare the WER between a transformer model with LM and without LM. 
I found I can simply do it by using the "valid_search" instead of "test_search" (specified in yaml file), 
then I can perform my test without LM.  (modified in train.py)

However, I encounter another question:
I found the input of transformer LM is the transformer encoder out, but not the transformer decoder out (predict). 
May I use the transformer decoder out (predict) as transformerLM input? 
" self.hparams.test_search(**enc_out**.detach(), wav_lens) "

Anyone has ideas about this?

**Reply by TParcollet** [reply]
It's because the decoder encapsulates the decoder.

**Reply by ccyang123** [reply]
@TParcollet 
Do you mean this is because the LM is a decoder, LM needs to decode the encoder output instead of a decoder output?

**Comment by ccyang123** [comment]
@TParcollet and @anautsch , May I ask another question?

When we evaluate the WER, I found we don't use the output of the transformer decoder as prediction results, but we use the LM or CTC outputs as the prediction results to evaluate the WER.  Is it correct? 

The outputs of transformer decoder are only used for training loss calculation. 
(It sounds a little strange.  )

Thank you for your advice.

**Reply by ccyang123** [reply]
After I read the python code of valid_search and test_search, I found how the WER is obtained.  FYI
